File: PongOnlineServer/PongOnlineServer/PongOnlineServer/Packet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PongOnlineServer {

    public enum Command { EnterGame, Disconnect, Move };
    public enum Props { Null, Success, Error, Up, Down };

    class Packet {

        private Command _command;

        public Command Command {
            get { return this._command; }
        }

        private Props _props;

        public Props Props {
            get { return this._props; }
        }

        public Packet(Command command, Props props) {
            this._command = command;
            this._props = props;
        }

        public Packet(byte[] streamData) {

            // (Lua-socket treats the hexadecimal value that is being sent as a char value, so we re-convert it to the numeric value).
            byte packetData = (byte)(streamData[0] - '0');

            int commandIndex = packetData >> 4;
            this._command = (Command)Enum.GetValues(Command.GetType()).GetValue(commandIndex);
            int propsIndex = packetData & 0x0F;
            this._props = (Props)Enum.GetValues(Props.GetType()).GetValue(propsIndex);
        }

        public static byte GeneratePacketByteData(Packet packet) {
            byte packetData = 0x00;
            packetData |= (byte)(packet.Command);
            packetData <<= sizeof(byte) / 2;
            packetData |= (byte)(packet.Props);
            return packetData;
        }
    }
}

File: PongOnlineClient/Networking/Packet.lua
-- Packet.ua

local Packet = {}
Packet.__index = Packet

-- Static definitions
Packet.Commands = {
  enterGame = 0x0,
  disconnect = 0x1,
  move = 0x2
}

-- TODO: Generate the key-sets by code

Packet.CommandsKeySet = {
  "enterGame",
  "disconnect",
  "move"
}

Packet.Props = {
  null = 0x0,
  success = 0x1,
  error = 0x2,
  up = 0x3,
  down = 0x4
}

-- TODO: Generate the key-sets by code

Packet.PropsKeySet = {
  "null",
  "success",
  "error",
  "up",
  "down"
}

function Packet.generatePacket(command, props)
  props = props or Packet.Props.null
  local packetData = 0x00
  packetData = ((command * 2^4) + props)
  return packetData
end

function Packet:new(byteData)
  local instance = {
    _command = Packet.Commands[Packet.CommandsKeySet[math.floor(byteData / 2^4) + 1]],
    _props   = Packet.Props[Packet.PropsKeySet[math.floor(byteData % 2^4) + 1]]
  }
  setmetatable(instance, self)
  return instance
end

return Packet
